processing/src/main/java/org/apache/druid/segment/DimensionDictionary.java

```diff
@@ -84,6 +84,19 @@ public T getValue(int id)
     }
   }
 
+  public void getValuesInto(int[] ids, T[] values)
+  {
+    lock.readLock().lock();
+    try {
+      for (int i = 0; i < ids.length; i++) {
```

Comment:
Since here you're copying the whole `ids` array to the input array without checking if the input `values` length matches the length of `ids`, I'm suggesting that this new method only accepts `ids` as parameter, and returns `T[]`

Reply:
Good point on checking array lengths match.

I did explore returning value only (that's my internal impl actually), but it requires wider changes. Specifically, since Java erases generics, it's impossible to do `new T[]`. So, we have to either do something like:
```
StringDimensionIndexer() constructor {
super(String.class);
}

// and then propagate this Class<T> to DimensionDictionary, and in DimensionDictionary, do a:

T[] out = (T[]) Array.newInstance(typeFromConstructor, keys.length);
```
To me, both are viable and work fine, this approach seems less invasive.

What do you suggest?

Reply:
Yes, handling generic in java is a little complicated. How about using `List<T>` as the returning type? This won't introduce any magic code.

Reply:
two paths here:
- explicit check solution in f4e4f9e5c0d0622e4f099e0668ae36d90a8a0f1d
- refactor to use reflection in f91f709a43462b2f7e8ab1339aedccbb10a10cfe

Reply:
#12073 is adding a `StringDimensionDictionary`, so this method might be able to be a bit cleaner in the future, https://github.com/apache/druid/pull/12073/files#diff-f8b1f0a09275a697d95c1d716e7c9f7650cdd2d7021497018df75e4635e4377cR28